### PR TITLE
Add Optimal Sugiyama Layout

### PR DESCRIPTION
**Pull request by oxinabox**
I am actually working more off this paper
https://ialab.it.monash.edu/~dwyer/papers/optimal-sankey-diagrams.pdf

so perhaps it should be a  ZarateLayout ?

TODO

- [x] Dummy Nodes
- [x] Permuting Order
- [x] Arranging
- [x] Weights for crossings
- [x] Do the perfomance constraint tweaks mentioned in the paper
- [x] Option to try multiple equally good orderings to see if any yeild better arrangement score.
- [x] Rewrite Arrangement subproblem from a QP into a SOCP and swap to ECOS
- [x] Vertically center the graph (right now the vertical position is basically arbitary, and thigns tend to be anywherre from 25-100, but really could be up in 10_000s or anywhere). Potentially this might cause stability issues (which centering in post would not fix) 

Also supports weights for crossings so should be perfect for Sankey
(cc @daschw)

### CODE REVIEW

**Comment by daschw** [comment]
Oooh, that looks amazing. Thanks for pinging!

**Comment by oxinabox** [comment]
sankey_3twos

@daschw for a Sankey diagram should the weights control the distance during arrangement, or just how hard to try for different overlapping line-crossings during ordering?

Currently it does both.
Which is why the 5 and 5 are lifted up in the diagram shown above.
The Tension from the 3->5 line is greater than that from the 3->6 line, so it pulls it up.

**Comment by oxinabox** [comment]
Oh wow my git settings are a total mess

**Comment by oxinabox** [comment]
I added checking for multiple solutions. It doesn't find many more but does find some more

**Comment by daschw** [comment]
> @daschw for a Sankey diagram should the weights control the distance during arrangement, or just how hard to try for different overlapping line-crossings during ordering?

Sorry, I missed this question. Looking at e.g. https://observablehq.com/@d3/sankey-diagram it does not seem like the weights impact the distance. It rather looks the maximum yaxis range is minimized.

**Comment by oxinabox** [comment]
>  Looking at e.g. https://observablehq.com/@d3/sankey-diagram it does not seem like the weights impact the distance. It rather looks the maximum yaxis range is minimized.

Cool, if we remove the weighting from that step then that is what we will have.
our notion of score for the distance step is the (currently weighted) sum of y-axis differences between connected vertexs. Which i think is what you are saying?

File: test/demos.jl [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        @plottest quick_plot_solve(layout, graph) ref_filename true tol
+    end
+end
+
+@testset "$layout Demos" for layout in (Zarate(),)# LayeredMinDistOne())
+    test_example(layout, :cross)
+    test_example(layout, :loop)
+    test_example(layout, :medium_pert)
+    test_example(layout, :sankey_3twos)
+    test_example(layout, :two_lines)
```
Comment (oxinabox) [inline test/demos.jl:37]:
```suggestion
    test_example(layout, :two_lines, 0.02)
```

File: test/demos.jl [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        w = 5*weights_mat[edge.src, edge.dst]
+        plot!(lxs, lys; linewidth=w, alpha=0.7, legend=false)
+    end
+end
+
+quick_plot_solve(layout, graph) = quick_plot(graph, solve_positions(layout, graph)...)
+
+@testset "quick_plot" begin
+    ref_filename =  joinpath(@__DIR__, "references", "test_utils", "$quick_plot.png")
+    @plottest quick_plot(SimpleDiGraph(Edge.([1=>2, 2=>3])), [1,2,5], [1,2,3]) ref_filename true 0.0001
```
Comment (oxinabox) [inline test/demos.jl:20]:
```suggestion
    @plottest quick_plot(SimpleDiGraph(Edge.([1=>2, 2=>3])), [1,2,5], [1,2,3]) ref_filename true 0.05
```

**Comment by oxinabox** [comment]
For some reason coveralls is failing to upload but tests are passing

**Comment by daschw** [comment]
I finally managed to continue work on this : https://github.com/daschw/SankeyPlots.jl

**Pull request merged**